drivers/modem/gsm_ppp.c

```diff
@@ -412,22 +418,53 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		NULL, 0, "AT+CGDCONT=1,\"IP\",\"" CONFIG_MODEM_GSM_APN "\"",
 		&gsm->sem_response, GSM_CMD_SETUP_TIMEOUT);
 
+attaching:
 	/* Don't initialize PPP until we're attached to packet service */
-	ret = modem_cmd_send_nolock(&gsm->context.iface,
-				    &gsm->context.cmd_handler,
-				    &check_attached_cmd, 1,
-				    "AT+CGATT?",
-				    &gsm->sem_response,
-				    GSM_CMD_SETUP_TIMEOUT);
+	ret = modem_cmd_send_nolock(
+		&gsm->context.iface, &gsm->context.cmd_handler,
+		&check_attached_cmd, 1, "AT+CGATT?", &gsm->sem_response,
+		GSM_CMD_SETUP_TIMEOUT);
 	if (ret < 0) {
+		/*
+		 * attach_retry not set        -> trigger 10 attach retries
+		 * attach_retry set            -> decrement and retry
+		 * attach_retry set, becomes 0 -> trigger full retry
+		 *
+		 * In total, modem gets 60 seconds to ack AT+CGATT which should
+		 * be fine for most RATs but possibly a making num retries
+		 * Kconfigurable would be best eventually.
+		 */
+		if (!attach_retry) {
+			attach_retry = 10;
+		} else {
+			attach_retry--;
+		}
+
 		LOG_DBG("Not attached, %s", "retrying...");
 		(void)k_delayed_work_submit(&gsm->gsm_configure_work,
-					    K_SECONDS(1));
+					    K_SECONDS(3));
 		return;
 	}
 
+	/* Clear attach_retry */
+	attach_retry = 0;
+
+#if defined(CONFIG_BOARD_WOLFENSTEIN)
+	/* For Quectel, dump the network info (LTE/WCDMA/etc) for debugging */
+	(void)modem_cmd_send_nolock(
+		&gsm->context.iface, &gsm->context.cmd_handler, NULL, 0,
+		"AT+QNWINFO", &gsm->sem_response, K_SECONDS(2));
+#endif
+
+	/* Ensure PDP context is activated. This is likely a NOP for most
+	 * modem/SIM combos, but it is required for certain setups.
+	 */
+	(void)modem_cmd_send_nolock(
+		&gsm->context.iface, &gsm->context.cmd_handler,
+		NULL, 0, "AT+CGACT=1,1",
+		&gsm->sem_response, GSM_CMD_SETUP_TIMEOUT);
 
```

Comment:
Can we do AT+CGACT=1,1 on other modems, or is this a Quectel only thing?

Reply:
It should be harmless but on SIMCOM the whole connection sequence breaks for some reason.